tools/queuebench/bench.go

```diff
@@ -0,0 +1,149 @@
+)
+
+func newBench(cfg *config) *bench {
+	return &bench{
+		cfg: *cfg,
+	}
+}
+
+type bench struct {
+	cfg config
+
+	manager  *kafka.Manager
+	Producer *kafka.Producer
+	Consumer *kafka.Consumer
+}
+
+func (b *bench) Setup(ctx context.Context, kafkaCommonCfg kafka.CommonConfig) {
+	log.Println("prep kafka.Manager")
+	mCfg := kafkaCommonCfg
+	mCfg.Logger = mCfg.Logger.With(zap.String("role", "manager"))
+	mngr, err := newMngr(mCfg, b.cfg.brokers)
+	if err != nil {
+		log.Fatal(fmt.Sprintf("cannot create kafka manager: %s", err))
+	}
+
+	b.manager = mngr
+	if err := mngr.Healthy(ctx); err != nil {
+
+		log.Fatal(fmt.Sprintf("kafka cluster is not healthy (%s): %s", b.cfg.brokers, err))
+	}
+
+	log.Println("cluster confirmed healthy")
+
+	log.Println("creating benchmark kafka topics")
+	if err := createTopics(ctx, mngr, b.cfg.partitions, b.cfg.topics); err != nil {
+		log.Fatal("cannot create topics: %w", err)
+	}
+
+	pCfg := kafkaCommonCfg
+	pCfg.Logger = pCfg.Logger.With(zap.String("role", "producer"))
+	p, err := createProducer(pCfg)
+	if err != nil {
+		log.Fatalf("cannot create kafka producer: %s", err)
+	}
+
+	b.Producer = p
+
+	cCfg := kafkaCommonCfg
+	cCfg.Logger = cCfg.Logger.With(zap.String("role", "consumer"))
+	c, err := createConsumer(cCfg, b.cfg.topics)
+	if err != nil {
+		log.Fatal(fmt.Sprintf("cannot create consumer: %s", err))
+	}
+
+	b.Consumer = c
+}
+
+func (b *bench) Teardown(ctx context.Context) {
+	deleteTopics(ctx, b.manager, b.cfg.topics)
+	b.manager.Close()
+}
+
+func newMngr(commoncfg kafka.CommonConfig, brokers []string) (*kafka.Manager, error) {
+	commoncfg.ClientID = fmt.Sprintf("%s-manager", app)
+
+	cfg := kafka.ManagerConfig{
+		CommonConfig: commoncfg,
+	}
+
+	return kafka.NewManager(cfg)
+}
+
+func createTopics(ctx context.Context, mngr *kafka.Manager, partitions int, topics []apmqueue.Topic) error {
+	cfg := kafka.TopicCreatorConfig{
+		PartitionCount: partitions,
+	}
+
+	creator, err := mngr.NewTopicCreator(cfg)
+	if err != nil {
+		return fmt.Errorf("cannot instantiate topic creator: %w", err)
+	}
+
+	for _, topic := range topics {
+		err = creator.CreateTopics(ctx, topic)
+		if err != nil {
+			return fmt.Errorf("cannot create topics: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func deleteTopics(ctx context.Context, mngr *kafka.Manager, topics []apmqueue.Topic) {
+	log.Println("deleting benchmark kafka topics")
+	err := mngr.DeleteTopics(ctx, topics...)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func createProducer(commoncfg kafka.CommonConfig) (*kafka.Producer, error) {
+	commoncfg.ClientID = fmt.Sprintf("%s-producer", app)
+
+	return kafka.NewProducer(kafka.ProducerConfig{
+		CommonConfig: commoncfg,
+	})
+}
+
+func createConsumer(commoncfg kafka.CommonConfig, topics []apmqueue.Topic) (*kafka.Consumer, error) {
+	commoncfg.ClientID = fmt.Sprintf("%s-consumer", app)
+
+	return kafka.NewConsumer(kafka.ConsumerConfig{
+		CommonConfig: commoncfg,
+
+		Topics:    topics,
+		GroupID:   "zero",
```

Comment:
Why "zero"?

Reply:
I didn't want to repeat "consumer" again and this is an uncommon value that I'm using to search for during testing. I haven't spent much time thinking about this as I was focused on the overall functionality. Open to suggestions for a better value.

Reply:
I think I misunderstood some implications of `GroupID`. To prevent clients to try reading from past reached offsets due to belonging to the same group (as the cluster itself may not be recreated between runs) I changed it to be unique for the single run and time based, which should prevent conflicts.

See [`8ada4de` (#256)](https://github.com/elastic/apm-queue/pull/256/commits/8ada4ded9374bfea7a6e67dd6c2e7a6e23129eb4#diff-ab6dc1b362b05def35e2d65c689ac6234c2a73455016306e7af1ad54d94fc1ceR118)

Reply:
The topic name is unique per run, right? So I don't think it's necessary to also make this ID unique. Anyway, looks better now, thanks.